File: services/api/app/api/routes/outings.py
"""Private group outing, timeline, and invitation endpoints."""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Header, Request, Response, status

from app.api.deps import Actor, get_actor, get_repository
from app.api.repository import ApiRepository
from app.api.schemas import (
    ErrorResponse,
    ItineraryPreviewRequest,
    ItineraryRequest,
    OutingCheckinListResponse,
    OutingCreateRequest,
    OutingInviteAcceptResponse,
    OutingInviteCreateRequest,
    OutingInviteResponse,
    OutingListResponse,
    OutingResponse,
    OutingTimelineRequest,
    StopCheckinResponse,
)
from app.api.search_rate_limit import FixedWindowLimiter
from app.api.service import ApiService

router = APIRouter(tags=["outings"])
ERRORS = {
    403: {"model": ErrorResponse},
    404: {"model": ErrorResponse},
    409: {"model": ErrorResponse},
    422: {"model": ErrorResponse},
}


def get_itinerary_limiter(request: Request) -> FixedWindowLimiter:
    return request.app.state.itinerary_limiter


@router.post("/outings/{outing_id}/itinerary/preview", responses=ERRORS)
def preview_outing_itinerary(
    outing_id: UUID,
    request: ItineraryPreviewRequest,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
    limiter: Annotated[FixedWindowLimiter, Depends(get_itinerary_limiter)],
    response: Response,
) -> dict:
    limiter.check(actor.id)
    response.headers["Cache-Control"] = "no-store"
    return ApiService(repository).preview_outing_itinerary(outing_id, request, actor)


@router.put(
    "/outings/{outing_id}/itinerary", response_model=OutingResponse, responses=ERRORS
)
def replace_outing_itinerary(
    outing_id: UUID,
    request: ItineraryRequest,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
    idempotency_key: Annotated[str, Header(min_length=1, max_length=255)],
    http: Request,
    response: Response,
) -> OutingResponse | Response:
    response.headers["Cache-Control"] = "no-store"
    replay = http.scope.get("itinerary_authorized_replay")
    if replay is not None:
        ApiService(repository).authorize_outing_itinerary(outing_id, actor)
        return Response(
            content=replay.body,
            status_code=replay.status_code,
            media_type=replay.media_type,
            headers={"Idempotency-Replayed": "true", "Cache-Control": "no-store"},
        )
    return ApiService(repository).replace_outing_itinerary(outing_id, request, actor)


@router.post(
    "/contexts/{context_id}/outings",
    response_model=OutingResponse,
    status_code=status.HTTP_201_CREATED,
    responses=ERRORS,
)
def create_outing(
    context_id: UUID,
    request: OutingCreateRequest,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> OutingResponse:
    return ApiService(repository).create_outing(context_id, request, actor)


@router.get(
    "/contexts/{context_id}/outings",
    response_model=OutingListResponse,
    responses=ERRORS,
)
def list_context_outings(
    context_id: UUID,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> OutingListResponse:
    return ApiService(repository).list_context_outings(context_id, actor)


@router.put(
    "/outings/{outing_id}/timeline",
    response_model=OutingResponse,
    responses=ERRORS,
)
def replace_outing_timeline(
    outing_id: UUID,
    request: OutingTimelineRequest,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> OutingResponse:
    return ApiService(repository).replace_outing_timeline(outing_id, request, actor)


@router.post(
    "/outing-stops/{stop_id}/checkins",
    response_model=StopCheckinResponse,
    status_code=status.HTTP_201_CREATED,
    responses=ERRORS,
)
def check_in_to_stop(
    stop_id: UUID,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> StopCheckinResponse:
    """F46. Record that the actor reached this stop.

    There is no request body. The only things a check-in records are who
    pressed it and when, both of which the server already knows -- a body
    would be a place for a coordinate to arrive.
    """
    return ApiService(repository).check_in_to_stop(stop_id, actor)


@router.get(
    "/outings/{outing_id}/checkins",
    response_model=OutingCheckinListResponse,
    responses=ERRORS,
)
def list_outing_checkins(
    outing_id: UUID,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> OutingCheckinListResponse:
    return ApiService(repository).list_outing_checkins(outing_id, actor)


@router.post(
    "/outings/{outing_id}/invites",
    response_model=OutingInviteResponse,
    status_code=status.HTTP_201_CREATED,
    responses=ERRORS,
)
def create_outing_invite(
    outing_id: UUID,
    request: OutingInviteCreateRequest,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> OutingInviteResponse:
    return ApiService(repository).create_outing_invite(outing_id, request, actor)


@router.post(
    "/outings/{outing_id}/invites/{invite_id}/revoke",
    response_model=OutingInviteResponse,
    responses=ERRORS,
)
def revoke_outing_invite(
    outing_id: UUID,
    invite_id: UUID,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> OutingInviteResponse:
    return ApiService(repository).revoke_outing_invite(outing_id, invite_id, actor)


@router.post(
    "/outings/{outing_id}/invites/{invite_id}/rotate",
    response_model=OutingInviteResponse,
    responses=ERRORS,
)
def rotate_outing_invite(
    outing_id: UUID,
    invite_id: UUID,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> OutingInviteResponse:
    # How a member who lost their phone gets back in. Not a second invitation:
    # the partial unique index allows one named row per person per outing, so
    # the way back is a new secret on the row that already names them.
    return ApiService(repository).rotate_outing_invite_secret(
        outing_id, invite_id, actor
    )


@router.post(
    "/outing-invites/{token}/accept",
    response_model=OutingInviteAcceptResponse,
    responses=ERRORS,
)
def accept_outing_invite(
    token: str,
    actor: Annotated[Actor, Depends(get_actor)],
    repository: Annotated[ApiRepository, Depends(get_repository)],
) -> OutingInviteAcceptResponse:
    return ApiService(repository).accept_outing_invite(token, actor)

File: services/core/internal/routes/outings.go
package routes

import (
	"context"
	"errors"
	"os"
	"strconv"
	"strings"
	"time"

	"mobile/services/core/internal/auth"
	"mobile/services/core/internal/domain/itinerary"
	"mobile/services/core/internal/domain/outingsteps"
	"mobile/services/core/internal/httpapi/endpoint"
	"mobile/services/core/internal/idem"
	"mobile/services/core/internal/routing"
	guestweb "mobile/services/core/internal/web/guest"
)

// The eleven routes of app/api/routes/outings.py. Each handler reads what
// pyval validated, hands it to the matching outingsteps method over a Store
// that is this request's transaction, and renders what comes back in the
// response model's field order. Preview is the one that may open a socket:
// outingsteps stops at the draft, and itinerary.Build calls the configured
// Valhalla (or answers unavailable when there is none).

func outingActor(call *endpoint.Call) outingsteps.Actor {
	return outingsteps.Actor{ID: call.Actor.ID, Roles: call.Actor.Roles}
}

func outingRefusal(err error) error {
	var refused *outingsteps.Refusal
	if errors.As(err, &refused) {
		return endpoint.Refuse(refused.Status, refused.Code, refused.Detail)
	}
	return err
}

type inviteSecrets struct{}

func (inviteSecrets) NewInviteToken() (string, []byte, error) {
	raw, err := mintGuestToken()
	if err != nil {
		return "", nil, err
	}
	return raw, auth.TokenDigest(raw), nil
}

func spendItineraryLimiter(call *endpoint.Call) error {
	if call.Limits == nil || call.Limits.ItineraryLimiter == nil {
		return errNoLimits
	}
	decision := call.Limits.ItineraryLimiter.Check(call.Actor.ID)
	if !decision.Allowed {
		return &endpoint.Refusal{Problem: decision.Problem}
	}
	return nil
}

func previewItinerary(draft itinerary.Draft) (itinerary.Preview, error) {
	router := routing.Configured()
	env := itinerary.Env{
		SuggestionsEnabled: os.Getenv("MOBILE_JOURNEY_SUGGESTIONS_ENABLED") != "0",
		CapacityAvailable:  true,
	}
	if router != nil {
		if !routing.TrySlot() {
			env.CapacityAvailable = false
		} else {
			defer routing.ReleaseSlot()
		}
	}
	return itinerary.Build(draft, router, env)
}

func itineraryReplay(stored idem.StoredResponse) endpoint.Reply {
	headers := [][2]string{
		{idem.ReplayHeaderName, "true"},
		{"Cache-Control", "no-store"},
	}
	if !(stored.Status < 200 || stored.Status == 204 || stored.Status == 304) {
		headers = append(headers, [2]string{"content-length", strconv.Itoa(len(stored.Body))})
	}
	if stored.MediaType != nil && *stored.MediaType != "" {
		media := *stored.MediaType
		if strings.HasPrefix(media, "text/") && !strings.Contains(strings.ToLower(media), "charset=") {
			media += "; charset=utf-8"
		}
		headers = append(headers, [2]string{"content-type", media})
	}
	return endpoint.Reply{Raw: &guestweb.Response{Status: stored.Status, Headers: headers, Body: stored.Body}}
}

func previewOutingItinerary() Route {
	return Route{ID: "POST /outings/{outing_id}/itinerary/preview", Status: 200, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		outingID, err := pathUUID(call, "outing_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		body, err := bodyModel(call, "request")
		if err != nil {
			return endpoint.Reply{}, err
		}
		request, err := itineraryPreviewRequest(body)
		if err != nil {
			return endpoint.Reply{}, err
		}
		if err := spendItineraryLimiter(call); err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		draft, err := outingsteps.PreviewOutingItinerary(store, outingID, request, outingActor(call))
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		preview, err := previewItinerary(draft)
		if err != nil {
			return endpoint.Reply{}, err
		}
		return endpoint.Reply{Body: wireItineraryPreview(preview), Headers: cacheControlNoStore}, nil
	}}
}

func replaceOutingItinerary() Route {
	return Route{ID: "PUT /outings/{outing_id}/itinerary", Status: 200, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		outingID, err := pathUUID(call, "outing_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		if stored, ok := idem.AuthorizedReplay(call.Request.Context()); ok {
			if _, err := outingsteps.AuthorizeOutingItinerary(store, outingID, outingActor(call)); err != nil {
				return endpoint.Reply{}, outingRefusal(err)
			}
			return itineraryReplay(stored), nil
		}
		body, err := bodyModel(call, "request")
		if err != nil {
			return endpoint.Reply{}, err
		}
		request, err := itineraryRequest(body)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.ReplaceOutingItinerary(store, outingID, request, outingActor(call))
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireOutingView(view), Headers: cacheControlNoStore}, nil
	}}
}

func createOuting() Route {
	return Route{ID: "POST /contexts/{context_id}/outings", Status: 201, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		contextID, err := pathUUID(call, "context_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		body, err := bodyModel(call, "request")
		if err != nil {
			return endpoint.Reply{}, err
		}
		request, err := outingCreateRequest(body)
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.CreateOuting(store, contextID, request, outingActor(call), time.Now().UTC())
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireOutingView(view)}, nil
	}}
}

func listContextOutings() Route {
	return Route{ID: "GET /contexts/{context_id}/outings", Status: 200, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		contextID, err := pathUUID(call, "context_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.ListContextOutings(store, contextID, outingActor(call))
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireOutingList(view)}, nil
	}}
}

func replaceOutingTimeline() Route {
	return Route{ID: "PUT /outings/{outing_id}/timeline", Status: 200, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		outingID, err := pathUUID(call, "outing_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		body, err := bodyModel(call, "request")
		if err != nil {
			return endpoint.Reply{}, err
		}
		request, err := timelineRequest(body)
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.ReplaceOutingTimeline(store, outingID, request, outingActor(call))
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireOutingView(view)}, nil
	}}
}

func checkInToStop() Route {
	return Route{ID: "POST /outing-stops/{stop_id}/checkins", Status: 201, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		stopID, err := pathUUID(call, "stop_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.CheckInToStop(store, stopID, outingActor(call), time.Now().UTC())
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireCheckinView(view)}, nil
	}}
}

func listOutingCheckins() Route {
	return Route{ID: "GET /outings/{outing_id}/checkins", Status: 200, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		outingID, err := pathUUID(call, "outing_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.ListOutingCheckins(store, outingID, outingActor(call))
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireCheckinList(view)}, nil
	}}
}

func createOutingInvite() Route {
	return Route{ID: "POST /outings/{outing_id}/invites", Status: 201, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		outingID, err := pathUUID(call, "outing_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		body, err := bodyModel(call, "request")
		if err != nil {
			return endpoint.Reply{}, err
		}
		request, err := inviteCreateRequest(body)
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.CreateOutingInvite(store, inviteSecrets{}, outingID, request, outingActor(call), time.Now().UTC())
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireInviteView(view)}, nil
	}}
}

func revokeOutingInvite() Route {
	return Route{ID: "POST /outings/{outing_id}/invites/{invite_id}/revoke", Status: 200, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		outingID, err := pathUUID(call, "outing_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		inviteID, err := pathUUID(call, "invite_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.RevokeOutingInvite(store, outingID, inviteID, outingActor(call), time.Now().UTC())
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireInviteView(view)}, nil
	}}
}

func rotateOutingInvite() Route {
	return Route{ID: "POST /outings/{outing_id}/invites/{invite_id}/rotate", Status: 200, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		outingID, err := pathUUID(call, "outing_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		inviteID, err := pathUUID(call, "invite_id")
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.RotateOutingInviteSecret(store, inviteSecrets{}, outingID, inviteID, outingActor(call), time.Now().UTC())
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireInviteView(view)}, nil
	}}
}

func acceptOutingInvite() Route {
	return Route{ID: "POST /outing-invites/{token}/accept", Status: 200, Serve: func(ctx context.Context, call *endpoint.Call) (endpoint.Reply, error) {
		token, err := stringParam(call, "token")
		if err != nil {
			return endpoint.Reply{}, err
		}
		store, err := newOutingStore(ctx, call)
		if err != nil {
			return endpoint.Reply{}, err
		}
		view, err := outingsteps.AcceptOutingInvite(store, auth.TokenDigest(token), outingActor(call), time.Now().UTC())
		if err != nil {
			return endpoint.Reply{}, outingRefusal(err)
		}
		return endpoint.Reply{Body: wireInviteAccept(view)}, nil
	}}
}
